src/lib/Caching/MPISharedStorageCache.cpp

```diff
@@ -0,0 +1,333 @@
+//===-- MPISharedStorageCache.cpp -- MPI shared storage cache impl --===//
+//
+// Part of the Proteus Project, under the Apache License v2.0 with LLVM
+// Exceptions. See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+//===----------------------------------------------------------------------===//
+
+#include <cstring>
+#include <filesystem>
+#include <limits>
+
+#include <mpi.h>
+
+#include <llvm/ADT/StringRef.h>
+#include <llvm/Support/MemoryBuffer.h>
+
+#include "proteus/Caching/MPISharedStorageCache.hpp"
+#include "proteus/Config.hpp"
+#include "proteus/Error.h"
+#include "proteus/Logger.hpp"
+#include "proteus/TimeTracing.hpp"
+#include "proteus/Utils.h"
+
+namespace proteus {
+
+using namespace llvm;
+namespace {
+constexpr int MaxRequestsPerCall = 5;
+} // namespace
+
+struct UnpackedMessage {
+  HashT Hash;
+  std::vector<char> Data;
+  bool IsDynLib;
+};
+
+static UnpackedMessage unpackMessage(const std::vector<char> &Buffer) {
```

Comment:
There is `MPI_pack/MPI_unpack` which should more robust and optimized than our own handrolled approach